build: Enable -Wsuggest-override if available

hebasto opened the pull request:
From GCC [docs](https://gcc.gnu.org/onlinedocs/gcc-9.2.0/gcc/Warning-Options.html):
> `-Wsuggest-override`
> Warn about overriding virtual functions that are not marked with the override keyword.

~This PR is based on #16722 (the first commit).~ See: https://github.com/bitcoin/bitcoin/pull/16722#issuecomment-584111086

practicalswift commented:
Concept ACK -- consistent use of `override` prevents bugs

@hebasto Is this change complete in the sense that it addresses all instances of missing `override` in the code?

Note that `clang`'s `-Winconsistent-missing-override` only checks for inconsistent `override` usage within a class. `gcc`:s `-Wsuggest-override` checks for missing overrides in general. `clang-tidy`:s `modernize-use-override` is helpful too.

hebasto commented:
@practicalswift 
> Note that `clang`'s `-Winconsistent-missing-override` only checks for inconsistent `override` usage **within a class**.

I did know that. Thank you.

hebasto commented:
@practicalswift 
> Is this change complete in the sense that it addresses all instances of missing `override` in the code?

Yes, it is (with the latest push).

practicalswift commented:
ACK 6a8d284bf7595bd200cb27973a262c929cb72e14 -- diff looks correct

Thanks for making the change complete across the codebase.

I think we should consider building with `-Werror=suggest-override` by default. But that is perhaps out of the scope this PR :-)

promag commented:
Concept ACK.

> But that is perhaps out of the scope this

Why? Otherwise this may be necessary again right?

hebasto commented:
@practicalswift  @promag Thank you for your reviews.

> I think we should consider building with `-Werror=suggest-override` by default.

How to skip warnings on `leveldb` subtree and `qt/paymentrequest.pb.h`?

practicalswift commented:
> > But that is perhaps out of the scope this
> 
> Why? Otherwise this may be necessary again right?

True! If @hebasto wants to incorporate `-Werror=suggest-override` as part of this PR I'm first in line to give a warm ACK :-)

practicalswift commented:
@hebasto 

> How to skip warnings on leveldb subtree and qt/paymentrequest.pb.h?

Could you make use of `-isystem`? See examples in @Empact's PRs #14920 and #15112.

hebasto commented:
@practicalswift 
> Could you make use of `-isystem`?

@theuni https://github.com/bitcoin/bitcoin/pull/14920#issuecomment-452505849
> I'm really not a fan of disabling warnings for 3rd party headers...

I tend to agree with @theuni, so I'm going to leave this PR as it is for now.

hebasto commented:
@practicalswift @promag 
>> Why? Otherwise this may be necessary again right?
>
> True! If @hebasto wants to incorporate -Werror=suggest-override as part of this PR I'm first in line to give a warm ACK :-)

The latest push incorporates `-Wsuggest-override` as a compiler option.

PR description has been updated.

hebasto commented:
Fixed MSVC build.

hebasto commented:
Fixed protobuf stuff.

hebasto commented:
@practicalswift Would you mind re-reviewing this PR?

hebasto commented:
Rebased.

laanwj commented:
Concept ACK.

Not convinced about `--enable-wleveldb`. Unless the warnings are purely silly, I think we should try to get these problems solved upstream not hide them.

> I think we should consider building with -Werror=suggest-override by default. But that is perhaps out of the scope this PR :-)

NACK on WError of any kind by default. Feel free to enable it locally as a developer, but it is frustrating for packagers and people just trying to build the code, because changes in gcc version, dependency libraries and such frequently change the set of warnings.

hebasto commented:
> Not convinced about `--enable-wleveldb`.

It seems there is a commit https://github.com/google/leveldb/commit/28e6d238be73e743c963fc0a26395b783a7565e2 to upstream related to `-Wsuggest-override`.

> Unless the warnings are purely silly, I think we should try to get these problems solved upstream not hide them.

E.g., #15539.

laanwj commented:
> It seems there is a commit google/leveldb@28e6d23 to upstream related to -Wsuggest-override.

Nice, let's pull that in. It might be time to do a leveldb update from upstream anyhow for 0.20.

> E.g., #15539.

Yes, would agree those are silly. I guess we could disable specific warnings for leveldb, if there's no change of them being solved, I just dislike the idea of a blanket disable.

hebasto commented:
Rebased.

practicalswift commented:
ACK 46047bc4e2bb9063e35fb56b17236eead4052ea8 -- diff looks correct

Thanks for keeping this PR rebased!

hebasto commented:
Rebased and dropped a commit from #16722; see: https://github.com/bitcoin/bitcoin/pull/16722#issuecomment-584111086.

There are 3 warnings from the `src/leveldb` subtree only:
```
$ gcc --version | grep gcc
gcc (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0
$ ./autogen.sh
$ ./configure
$ make -j 4 > /dev/null
/usr/bin/ar: `u' modifier ignored since `D' is the default (see `U')
ar: `u' modifier ignored since `D' is the default (see `U')
leveldb/db/c.cc: In function ‘leveldb_filterpolicy_t* leveldb_filterpolicy_create_bloom(int)’:
leveldb/db/c.cc:474:17: warning: ‘virtual const char* leveldb_filterpolicy_create_bloom(int)::Wrapper::Name() const’ can be marked override [-Wsuggest-override]
     const char* Name() const { return rep_->Name(); }
                 ^~~~
leveldb/db/c.cc:475:10: warning: ‘virtual void leveldb_filterpolicy_create_bloom(int)::Wrapper::CreateFilter(const leveldb::Slice*, int, std::__cxx11::string*) const’ can be marked override [-Wsuggest-override]
     void CreateFilter(const Slice* keys, int n, std::string* dst) const {
          ^~~~~~~~~~~~
leveldb/db/c.cc:478:10: warning: ‘virtual bool leveldb_filterpolicy_create_bloom(int)::Wrapper::KeyMayMatch(const leveldb::Slice&, const leveldb::Slice&) const’ can be marked override [-Wsuggest-override]
     bool KeyMayMatch(const Slice& key, const Slice& filter) const {
          ^~~~~~~~~~~
/usr/bin/ar: `u' modifier ignored since `D' is the default (see `U')
```

hebasto commented:
Rebased after #18034 has been merged.

practicalswift commented:
ACK ec530398dbb3138d26c4783f880e4efe439fc450 -- looks good: let's get this merged :)

Empact commented:
Code Review ACK https://github.com/bitcoin/bitcoin/pull/16710/commits/ec530398dbb3138d26c4783f880e4efe439fc450 - also confirmed moves out of Q_SLOTS were appropriate.

MarcoFalke commented on src/script/descriptor.cpp:
```diff
@@ -514,7 +514,7 @@ class AddressDescriptor final : public DescriptorImpl
 public:
     AddressDescriptor(CTxDestination destination) : DescriptorImpl({}, {}, "addr"), m_destination(std::move(destination)) {}
-    bool IsSolvable() const final { return false; }
+    bool IsSolvable() const override final { return false; }
```
This is final, but not virtual, so override is already implied. I am not sure why the compiler warns here?!

fanquake replied:
GCC (Debian 8.3.0-6) will warn about this:
```bash
  CXX      script/libbitcoin_common_a-descriptor.o
script/descriptor.cpp:517:10: warning: 'virtual bool {anonymous}::AddressDescriptor::IsSolvable() const' can be marked override [-Wsuggest-override]
     bool IsSolvable() const final { return false; }
          ^~~~~~~~~~
```

However `clang-tidy` with [modernize-use-override](https://clang.llvm.org/extra/clang-tidy/checks/modernize-use-override.html) will tell you that it, along with a bunch of the other overrides added here are redundant:
```bash
src/script/descriptor.cpp:378:10: warning: 'override' is redundant since the function is already declared 'final' [modernize-use-override]
    bool IsRange() const override final
         ^               ~~~~~~~~~
src/script/descriptor.cpp:414:17: warning: 'override' is redundant since the function is already declared 'final' [modernize-use-override]
    std::string ToString() const override final
                ^                ~~~~~~~~~
src/script/descriptor.cpp:421:10: warning: 'override' is redundant since the function is already declared 'final' [modernize-use-override]
    bool ToPrivateString(const SigningProvider& arg, std::string& out) const override final
         ^                                                                   ~~~~~~~~~
src/script/descriptor.cpp:480:10: warning: 'override' is redundant since the function is already declared 'final' [modernize-use-override]
    bool Expand(int pos, const SigningProvider& provider, std::vector<CScript>& output_scripts, FlatSigningProvider& out, std::vector<unsigned char>* cache = nullptr) const override final
         ^                                                                                                                                                                   ~~~~~~~~~
src/script/descriptor.cpp:485:10: warning: 'override' is redundant since the function is already declared 'final' [modernize-use-override]
    bool ExpandFromCache(int pos, const std::vector<unsigned char>& cache, std::vector<CScript>& output_scripts, FlatSigningProvider& out) const override final
         ^                                                                                                                                       ~~~~~~~~~
src/script/descriptor.cpp:491:10: warning: 'override' is redundant since the function is already declared 'final' [modernize-use-override]
    void ExpandPrivate(int pos, const SigningProvider& provider, FlatSigningProvider& out) const override final
         ^                                                                                       ~~~~~~~~~
src/script/descriptor.cpp:517:10: warning: 'override' is redundant since the function is already declared 'final' [modernize-use-override]
    bool IsSolvable() const override final { return false; }
         ^                  ~~~~~~~~~
src/script/descriptor.cpp:542:10: warning: 'override' is redundant since the function is already declared 'final' [modernize-use-override]
    bool IsSolvable() const override final { return false; }
         ^                  ~~~~~~~~~
```

hebasto replied:
> This is final, but not virtual, so override is already implied. I am not sure why the compiler warns here?!

This GCC behavior fixed for 9.2 ([GCC Bugzilla – Bug 78010](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=78010)).

hebasto replied:
> However `clang-tidy` with [modernize-use-override](https://clang.llvm.org/extra/clang-tidy/checks/modernize-use-override.html) will tell you that it, along with a bunch of the other overrides added here are redundant:

There is an [`AllowOverrideAndFinal`](https://clang.llvm.org/extra/clang-tidy/checks/modernize-use-override.html#cmdoption-arg-allowoverrideandfinal) option:
> If set to non-zero, this check will not diagnose `override` as redundant with `final`. This is useful when code will be compiled by a compiler with warning/error checking flags requiring `override` explicitly on overridden members, such as `gcc -Wsuggest-override`/`gcc -Werror=suggest-override`. Default is _0_.

hebasto replied:
Addressed in the latest [push](https://github.com/bitcoin/bitcoin/pull/16710#issuecomment-593608787).

MarcoFalke commented:
I am fine with the `src/qt` changes, but I am not sure if it make sense to "enforce" this with a warning that spits out misleading information on older and buggy compilers.

theuni commented:
It may make sense to use a bunch of test-cases to test whether or not we should enable this, and include those cases that are known to cause issues with buggy compilers. Then just say "if your compiler can't handle all test-cases correctly, we'll just turn the warnings off wholesale".

IIRC the c++xy m4 test macro takes a similar approach.

hebasto commented:
Updated ec530398dbb3138d26c4783f880e4efe439fc450 -> f33cc5446b3d8358fa274cfee0287b30f82ceaab ([pr16710.10](https://github.com/hebasto/bitcoin/commits/pr16710.10) -> [pr16710.11](https://github.com/hebasto/bitcoin/commits/pr16710.11), [compare](https://github.com/hebasto/bitcoin/compare/pr16710.10..pr16710.11)):

- removed ignoring of GCC "-Wsuggest-override" diagnostic for `leveldb/env.h` header as it is fixed in #17398 
- added ignoring of GCC "-Wsuggest-override" diagnostic for `db_cxx.h` header as it causes spammy warnings on Fedora 31 (GCC 9.2.1, `libdb4-cxx-devel.x86_64`::4.8.30-30.fc31)
- addressed [comments](https://github.com/bitcoin/bitcoin/pull/16710#discussion_r386139557) about final overriding functions

---

**GCC 7.4.0** output:
```
$ gcc --version | head -1
gcc (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0
$ make > /dev/null
/usr/bin/ar: `u' modifier ignored since `D' is the default (see `U')
ar: `u' modifier ignored since `D' is the default (see `U')
leveldb/db/c.cc: In function ‘leveldb_filterpolicy_t* leveldb_filterpolicy_create_bloom(int)’:
leveldb/db/c.cc:474:17: warning: ‘virtual const char* leveldb_filterpolicy_create_bloom(int)::Wrapper::Name() const’ can be marked override [-Wsuggest-override]
     const char* Name() const { return rep_->Name(); }
                 ^~~~
leveldb/db/c.cc:475:10: warning: ‘virtual void leveldb_filterpolicy_create_bloom(int)::Wrapper::CreateFilter(const leveldb::Slice*, int, std::__cxx11::string*) const’ can be marked override [-Wsuggest-override]
     void CreateFilter(const Slice* keys, int n, std::string* dst) const {
          ^~~~~~~~~~~~
leveldb/db/c.cc:478:10: warning: ‘virtual bool leveldb_filterpolicy_create_bloom(int)::Wrapper::KeyMayMatch(const leveldb::Slice&, const leveldb::Slice&) const’ can be marked override [-Wsuggest-override]
     bool KeyMayMatch(const Slice& key, const Slice& filter) const {
          ^~~~~~~~~~~
/usr/bin/ar: `u' modifier ignored since `D' is the default (see `U')
```

---


**GCC 9.2.1** output:
```
$ gcc --version | head -1
gcc (GCC) 9.2.1 20190827 (Red Hat 9.2.1-1)
$ qmake-qt5 -v | grep Qt
Using Qt version 5.13.2 in /usr/lib64
$ make > /dev/null 
In file included from /usr/include/string.h:495,
                 from ./serialize.h:19,
                 from ./netaddress.h:13,
                 from ./protocol.h:13,
                 from protocol.cpp:6:
In function ‘char* strncpy(char*, const char*, size_t)’,
    inlined from ‘CMessageHeader::CMessageHeader(const unsigned char (&)[4], const char*, unsigned int)’ at protocol.cpp:89:12:
/usr/include/bits/string_fortified.h:106:34: warning: ‘char* __builtin_strncpy(char*, const char*, long unsigned int)’ specified bound 12 equals destination size [-Wstringop-truncation]
  106 |   return __builtin___strncpy_chk (__dest, __src, __len, __bos (__dest));
      |          ~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
leveldb/db/c.cc: In function ‘leveldb_filterpolicy_t* leveldb_filterpolicy_create_bloom(int)’:
leveldb/db/c.cc:474:17: warning: ‘virtual const char* leveldb_filterpolicy_create_bloom(int)::Wrapper::Name() const’ can be marked override [-Wsuggest-override]
  474 |     const char* Name() const { return rep_->Name(); }
      |                 ^~~~
leveldb/db/c.cc:475:10: warning: ‘virtual void leveldb_filterpolicy_create_bloom(int)::Wrapper::CreateFilter(const leveldb::Slice*, int, std::string*) const’ can be marked override [-Wsuggest-override]
  475 |     void CreateFilter(const Slice* keys, int n, std::string* dst) const {
      |          ^~~~~~~~~~~~
leveldb/db/c.cc:478:10: warning: ‘virtual bool leveldb_filterpolicy_create_bloom(int)::Wrapper::KeyMayMatch(const leveldb::Slice&, const leveldb::Slice&) const’ can be marked override [-Wsuggest-override]
  478 |     bool KeyMayMatch(const Slice& key, const Slice& filter) const {
      |          ^~~~~~~~~~~
interfaces/chain.cpp: In member function ‘virtual std::unique_ptr<interfaces::Chain::Lock> interfaces::{anonymous}::ChainImpl::lock(bool)’:
interfaces/chain.cpp:250:25: warning: redundant move in return statement [-Wredundant-move]
  250 |         return std::move(result);
      |                ~~~~~~~~~^~~~~~~~
interfaces/chain.cpp:250:25: note: remove ‘std::move’ call

```

hebasto commented:
@theuni 
> It may make sense to use a bunch of test-cases to test whether or not we should enable this, and include those cases that are known to cause issues with buggy compilers. Then just say "if your compiler can't handle all test-cases correctly, we'll just turn the warnings off wholesale".
> 
> IIRC the c++xy m4 test macro takes a similar approach.

Yes, m4 macro could test if `-Wsuggest-override` is buggy for a particular GCC which is used for compiling.

But it seems overkill to "turn the warnings off wholesale". Especially considering that "buggy" GCC just fires some false-positive warnings, which could be easily suppressed in well-documented manner.

practicalswift commented:
ACK f33cc5446b3d8358fa274cfee0287b30f82ceaab

hebasto commented:
@MarcoFalke 
> I am fine with the `src/qt` changes, but I am not sure if it make sense to "enforce" this with a warning that spits out misleading information on older and buggy compilers.

If a false positive warning suppression approach seems too wordy, the last two commits could be dropped. In that case we will refuse benefits of `-Wsuggest-override`.

theuni commented:
> But it seems overkill to "turn the warnings off wholesale". Especially considering that "buggy" GCC just fires some false-positive warnings, which could be easily suppressed in well-documented manner.

If a warning flag is known to be buggy and it can be detected, we should turn it off. It's perfectly reasonable imo to say "If override warnings for this compiler are known to be broken, don't enable them".

hebasto commented:
Updated f33cc5446b3d8358fa274cfee0287b30f82ceaab -> d09c985ac4c7289875529da68f4a0ed1f04361f6 ([pr16710.11](https://github.com/hebasto/bitcoin/commits/pr16710.11) -> [pr16710.12](https://github.com/hebasto/bitcoin/commits/pr16710.12), [compare](https://github.com/hebasto/bitcoin/compare/pr16710.11..pr16710.12)):

- fixed MSVC build

practicalswift commented:
ACK d09c985ac4c7289875529da68f4a0ed1f04361f6

hebasto committed:
refactor: Add BerkeleyDatabaseVersion() function

hebasto committed:
build: Enable -Wsuggest-override

hebasto commented:
Rebased due to the conflict with #18204.

practicalswift commented:
ACK 2e9e8c8122c3c837e61f8b8b703a7807762ee550 -- patch looks correct

vasild commented:
ACK 2e9e8c812, compiles with Clang 11.0.0.

The more `-W...` the better :)

practicalswift commented:
Reviewers of this PR might be interested in reviewing #17895 which enables Clang's `-Wconditional-uninitialized` (warn on potential uninitialized reads).

hebasto commented:
Rebased due to the conflict with #18843.

MarcoFalke commented:
What do you think about splitting out the gui changes from the build system changes? They are the bulk of the changes, but the easiest to review. Mostly move-only + adding the override, whereas the other changes are refactoring walletdb and changing the build stystem.

hebasto commented:
@MarcoFalke 
> What do you think about splitting out the gui changes from the build system changes? They are the bulk of the changes, but the easiest to review. Mostly move-only + adding the override, whereas the other changes are refactoring walletdb and changing the build stystem.

Done in #18914.

practicalswift commented:
ACK a96adef72a837c7988ee45990a69d9b15f0b7ef1

vasild commented on configure.ac:
```diff
@@ -383,6 +383,7 @@ if test "x$CXXFLAGS_overridden" = "xno"; then
   AX_CHECK_COMPILE_FLAG([-Wunused-variable],[WARN_CXXFLAGS="$WARN_CXXFLAGS -Wunused-variable"],,[[$CXXFLAG_WERROR]])
   AX_CHECK_COMPILE_FLAG([-Wdate-time],[WARN_CXXFLAGS="$WARN_CXXFLAGS -Wdate-time"],,[[$CXXFLAG_WERROR]])
   AX_CHECK_COMPILE_FLAG([-Wconditional-uninitialized],[WARN_CXXFLAGS="$WARN_CXXFLAGS -Wconditional-uninitialized"],,[[$CXXFLAG_WERROR]])
+  AX_CHECK_COMPILE_FLAG([-Wsuggest-override],[WARN_CXXFLAGS="$WARN_CXXFLAGS -Wsuggest-override"],,[[$CXXFLAG_WERROR]])
```
Enabling `-Wsuggest-override` produces new warnings (with gcc 10):

```
.../src/leveldb/db/c.cc: In function 'leveldb_filterpolicy_t* leveldb_filterpolicy_create_bloom(int)':
.../src/leveldb/db/c.cc:474:17: warning: 'virtual const char* leveldb_filterpolicy_create_bloom(int)::Wrapper::Name() const' can be marked override [-Wsuggest-override]
  474 |     const char* Name() const { return rep_->Name(); }
      |                 ^~~~
.../src/leveldb/db/c.cc:475:10: warning: 'virtual void leveldb_filterpolicy_create_bloom(int)::Wrapper::CreateFilter(const leveldb::Slice*, int, std::string*) const' can be marked override [-Wsuggest-override]
  475 |     void CreateFilter(const Slice* keys, int n, std::string* dst) const {
      |          ^~~~~~~~~~~~
.../src/leveldb/db/c.cc:478:10: warning: 'virtual bool leveldb_filterpolicy_create_bloom(int)::Wrapper::KeyMayMatch(const leveldb::Slice&, const leveldb::Slice&) const' can be marked override [-Wsuggest-override]
  478 |     bool KeyMayMatch(const Slice& key, const Slice& filter) const {
      |          ^~~~~~~~~~~
```

I would suggest to not enable the new flag on LevelDB sources and to broaden the scope of `--enable-werror` like this:

```diff
diff --git i/configure.ac w/configure.ac
index 368541b0c..c35aa9343 100644
--- i/configure.ac
+++ w/configure.ac
@@ -364,12 +364,13 @@ if test "x$enable_werror" = "xyes"; then
   AX_CHECK_COMPILE_FLAG([-Werror=switch],[ERROR_CXXFLAGS="$ERROR_CXXFLAGS -Werror=switch"],,[[$CXXFLAG_WERROR]])
   AX_CHECK_COMPILE_FLAG([-Werror=thread-safety-analysis],[ERROR_CXXFLAGS="$ERROR_CXXFLAGS -Werror=thread-safety-analysis"],,[[$CXXFLAG_WERROR]])
   AX_CHECK_COMPILE_FLAG([-Werror=unused-variable],[ERROR_CXXFLAGS="$ERROR_CXXFLAGS -Werror=unused-variable"],,[[$CXXFLAG_WERROR]])
   AX_CHECK_COMPILE_FLAG([-Werror=date-time],[ERROR_CXXFLAGS="$ERROR_CXXFLAGS -Werror=date-time"],,[[$CXXFLAG_WERROR]])
   AX_CHECK_COMPILE_FLAG([-Werror=return-type],[ERROR_CXXFLAGS="$ERROR_CXXFLAGS -Werror=return-type"],,[[$CXXFLAG_WERROR]])
   AX_CHECK_COMPILE_FLAG([-Werror=conditional-uninitialized],[ERROR_CXXFLAGS="$ERROR_CXXFLAGS -Werror=conditional-uninitialized"],,[[$CXXFLAG_WERROR]])
+  AX_CHECK_COMPILE_FLAG([-Werror=suggest-override],[ERROR_CXXFLAGS="$ERROR_CXXFLAGS -Werror=suggest-override"],,[[$CXXFLAG_WERROR]])
 fi
 
 if test "x$CXXFLAGS_overridden" = "xno"; then
   AX_CHECK_COMPILE_FLAG([-Wall],[WARN_CXXFLAGS="$WARN_CXXFLAGS -Wall"],,[[$CXXFLAG_WERROR]])
   AX_CHECK_COMPILE_FLAG([-Wextra],[WARN_CXXFLAGS="$WARN_CXXFLAGS -Wextra"],,[[$CXXFLAG_WERROR]])
   AX_CHECK_COMPILE_FLAG([-Wgnu],[WARN_CXXFLAGS="$WARN_CXXFLAGS -Wgnu"],,[[$CXXFLAG_WERROR]])
diff --git i/src/Makefile.leveldb.include w/src/Makefile.leveldb.include
index 79ff72ca8..8a28f4f24 100644
--- i/src/Makefile.leveldb.include
+++ w/src/Makefile.leveldb.include
@@ -33,13 +33,13 @@ if TARGET_WINDOWS
 LEVELDB_CPPFLAGS_INT += -DLEVELDB_PLATFORM_WINDOWS -D_UNICODE -DUNICODE -D__USE_MINGW_ANSI_STDIO=1
 else
 LEVELDB_CPPFLAGS_INT += -DLEVELDB_PLATFORM_POSIX
 endif
 
 leveldb_libleveldb_a_CPPFLAGS = $(AM_CPPFLAGS) $(LEVELDB_CPPFLAGS_INT) $(LEVELDB_CPPFLAGS)
-leveldb_libleveldb_a_CXXFLAGS = $(filter-out -Wconditional-uninitialized -Werror=conditional-uninitialized, $(AM_CXXFLAGS)) $(PIE_FLAGS)
+leveldb_libleveldb_a_CXXFLAGS = $(filter-out -Wconditional-uninitialized -Werror=conditional-uninitialized -Wsuggest-override -Werror=suggest-override, $(AM_CXXFLAGS)) $(PIE_FLAGS)
 
 leveldb_libleveldb_a_SOURCES=
 leveldb_libleveldb_a_SOURCES += leveldb/port/port_stdcxx.h
 leveldb_libleveldb_a_SOURCES += leveldb/port/port.h
 leveldb_libleveldb_a_SOURCES += leveldb/port/thread_annotations.h
 leveldb_libleveldb_a_SOURCES += leveldb/include/leveldb/db.h
```

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/16710#issuecomment-626887028).

vasild reviewed (commented):
Approach ACK

hebasto commented:
Updated a96adef72a837c7988ee45990a69d9b15f0b7ef1 -> d44a13b4b2dac522e5f68ca5bdd9cd972ae6d5e9 ([pr16710.14](https://github.com/hebasto/bitcoin/commits/pr16710.14) -> [pr16710.15](https://github.com/hebasto/bitcoin/commits/pr16710.15)):

- rebased
- addressed @vasild's [comment](https://github.com/bitcoin/bitcoin/pull/16710#discussion_r423118228):
> I would suggest to not enable the new flag on LevelDB sources and to broaden the scope of `--enable-werror` like this...

I'd refrain from `-Werror=suggest-override` while we have buggy compilers like GCC < 9.2 (see the top commit).

MarcoFalke commented:
I don't think d44a13b4b2dac522e5f68ca5bdd9cd972ae6d5e9 is maintainable. How would devs with compilers that are not buggy figure out where to add the pragmas?

I think the only thing that can be done is enable the warning for compilers that support it and skip it if the compiler does not support it fully and correctly.

vasild commented:
ACK d44a13b4b

The current code contains just a few `final` methods so it is kind of ok to surround them with `#pragma suppress if gcc < 9.2`.

However, every addition of a new `final` method would either produce a warning with gcc < 9.2 or require to be also surrounded by the `#pragma` stuff. I wonder if it would be better to only conditionally enable `-Wsuggest-override` in `configure.ac` if the compiler is not gcc < 9.2?

hebasto commented:
Updated d44a13b4b2dac522e5f68ca5bdd9cd972ae6d5e9 -> ea6849fe2fb53bfd50e3805b45746a36208b0781 ([pr16710.15](https://github.com/hebasto/bitcoin/commits/pr16710.15) -> [pr16710.16](https://github.com/hebasto/bitcoin/commits/pr16710.16), [diff](https://github.com/hebasto/bitcoin/compare/pr16710.15..pr16710.16)):

- `-Wsuggest-override` and `-Werror=suggest-override` are enabled only for non-buggy GCC versions, i.e. 9.2+.

hebasto commented:
@theuni 
> > But it seems overkill to "turn the warnings off wholesale". Especially considering that "buggy" GCC just fires some false-positive warnings, which could be easily suppressed in well-documented manner.
> 
> If a warning flag is known to be buggy and it can be detected, we should turn it off. It's perfectly reasonable imo to say "If override warnings for this compiler are known to be broken, don't enable them".

Done in the [latest push](https://github.com/bitcoin/bitcoin/pull/16710#issuecomment-627216048).

vasild commented on configure.ac:
```diff
@@ -355,6 +355,24 @@ if test x$use_sanitizers != x; then
 
+dnl GCC<9.2 reports a redundant -Wsuggest-override warning on a 'final' method.
+dnl See https://gcc.gnu.org/bugzilla/show_bug.cgi?id=78010
+AC_MSG_CHECKING(for GCC compiler version < 9.2)
```
nit: just "... GCC version ..." is ok too (remove "compiler" because "GCC" means "GNU Compiler Collection")

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/16710#issuecomment-627245385).

vasild commented on configure.ac:
```diff
@@ -368,6 +386,9 @@ if test "x$enable_werror" = "xyes"; then
   AX_CHECK_COMPILE_FLAG([-Werror=conditional-uninitialized],[ERROR_CXXFLAGS="$ERROR_CXXFLAGS -Werror=conditional-uninitialized"],,[[$CXXFLAG_WERROR]])
   AX_CHECK_COMPILE_FLAG([-Werror=sign-compare],[ERROR_CXXFLAGS="$ERROR_CXXFLAGS -Werror=sign-compare"],,[[$CXXFLAG_WERROR]])
+  if test "x$have_gcc_suggest_override_bug" = "xno"; then
+    AX_CHECK_COMPILE_FLAG([-Werror=suggest-override],[WARN_CXXFLAGS="$ERROR_CXXFLAGS -Werror=suggest-override"],,[[$CXXFLAG_WERROR]])
```
`s/WARN_CXXFLAGS/ERROR_CXXFLAGS/`

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/16710#issuecomment-627245385).

hebasto commented:
Updated ea6849fe2fb53bfd50e3805b45746a36208b0781 -> e6211d8b54338418575ffb762860c7634e10c977 ([pr16710.16](https://github.com/hebasto/bitcoin/commits/pr16710.16) -> [pr16710.17](https://github.com/hebasto/bitcoin/commits/pr16710.17), [diff](https://github.com/hebasto/bitcoin/compare/pr16710.16..pr16710.17)):

- addressed @vasild's comments

vasild reviewed (approved):
ACK e6211d8b5

GCC 8.4.0:
```
checking for GCC version < 9.2... yes
configure: WARNING: GCC<9.2 reports a redundant -Wsuggest-override warning on a 'final' method, disabling -Wsuggest-override.
```
(and `-Wsuggest-override -Werror=suggest-override` are not in `CXXFLAGS`)


GCC 9.3.0:
```
checking for GCC version < 9.2... no
checking whether C++ compiler accepts -Werror=suggest-override... yes
checking whether C++ compiler accepts -Wsuggest-override... yes
```


Clang 11.0.0:
```
checking for GCC version < 9.2... no
checking whether C++ compiler accepts -Werror=suggest-override... no
checking whether C++ compiler accepts -Wsuggest-override... no
```

fanquake commented on configure.ac:
```diff
@@ -355,6 +355,24 @@ if test x$use_sanitizers != x; then
+  ],
+  [
+    AC_MSG_RESULT(yes)
+    AC_MSG_WARN([GCC<9.2 reports a redundant -Wsuggest-override warning on a 'final' method, disabling -Wsuggest-override.])
```
If we do keep this as is, let's drop this call to `AC_MSG_WARN` and the `AC_MSG_CHECKING` above. I don't think we need any extra, verbose out here. Especially since this will be the default case most people compiling going forward.

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/16710#issuecomment-627406029).

fanquake reviewed (changes requested):
How about something like this to replace the changes in `configure.ac`:
```diff
diff --git a/configure.ac b/configure.ac
index 7de6adb93..31e9c243f 100644
--- a/configure.ac
+++ b/configure.ac
@@ -368,6 +368,11 @@ if test "x$enable_werror" = "xyes"; then
   AX_CHECK_COMPILE_FLAG([-Werror=return-type],[ERROR_CXXFLAGS="$ERROR_CXXFLAGS -Werror=return-type"],,[[$CXXFLAG_WERROR]])
   AX_CHECK_COMPILE_FLAG([-Werror=conditional-uninitialized],[ERROR_CXXFLAGS="$ERROR_CXXFLAGS -Werror=conditional-uninitialized"],,[[$CXXFLAG_WERROR]])
   AX_CHECK_COMPILE_FLAG([-Werror=sign-compare],[ERROR_CXXFLAGS="$ERROR_CXXFLAGS -Werror=sign-compare"],,[[$CXXFLAG_WERROR]])
+
+  dnl -Wsuggest-override is  broken with GCC before 9.2
+  dnl https://gcc.gnu.org/bugzilla/show_bug.cgi?id=78010
+  AX_CHECK_COMPILE_FLAG([-Wsuggest-override],[WARN_CXXFLAGS="$WARN_CXXFLAGS -Wsuggest-override"],,[[$CXXFLAG_WERROR]],
+                       [AC_LANG_PROGRAM([[struct A { virtual void f(); };struct B : A { void f() final; };]])])
 fi
 
 if test "x$CXXFLAGS_overridden" = "xno"; then
@@ -385,6 +390,8 @@ if test "x$CXXFLAGS_overridden" = "xno"; then
   AX_CHECK_COMPILE_FLAG([-Wdate-time],[WARN_CXXFLAGS="$WARN_CXXFLAGS -Wdate-time"],,[[$CXXFLAG_WERROR]])
   AX_CHECK_COMPILE_FLAG([-Wconditional-uninitialized],[WARN_CXXFLAGS="$WARN_CXXFLAGS -Wconditional-uninitialized"],,[[$CXXFLAG_WERROR]])
   AX_CHECK_COMPILE_FLAG([-Wsign-compare],[WARN_CXXFLAGS="$WARN_CXXFLAGS -Wsign-compare"],,[[$CXXFLAG_WERROR]])
+  AX_CHECK_COMPILE_FLAG([-Wsuggest-override],[WARN_CXXFLAGS="$WARN_CXXFLAGS -Wsuggest-override"],,[[$CXXFLAG_WERROR]],
+                       [AC_LANG_PROGRAM([[struct A { virtual void f(); };struct B : A { void f() final; };]])])
 
   dnl Some compilers (gcc) ignore unknown -Wno-* options, but warn about all
   dnl unknown options if any other warning is produced. Test the -Wfoo case, and
```

This way the check stays self-contained, no need to test version macros, no additional configure output and we can just use the source from the bug report.

hebasto commented:
@fanquake Thank you for your review. TIL :)

Updated e6211d8b54338418575ffb762860c7634e10c977 -> 839add193b13c17a40f42ff69d973caeb800d3f2 ([pr16710.17](https://github.com/hebasto/bitcoin/commits/pr16710.17) -> [pr16710.18](https://github.com/hebasto/bitcoin/commits/pr16710.18), [diff](https://github.com/hebasto/bitcoin/compare/pr16710.17..pr16710.18)).

- addressed @fanquake's [comment](https://github.com/bitcoin/bitcoin/pull/16710#pullrequestreview-410029779)

practicalswift commented:
ACK 839add193b13c17a40f42ff69d973caeb800d3f2 assuming Travis is happy: patch looks correct

vasild commented:
ACK 839add193

Even better! I confirm it works as intended with GCC 8.4, 9.3 and Clang 11.0.

fanquake reviewed (approved):
ACK 839add193b13c17a40f42ff69d973caeb800d3f2
